include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -152,18 +152,77 @@ __pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first
     }
 }
 
+// TODO In C++20 we may try to use std::equality_comparable
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct is_equality_comparable : std::false_type
+{
+};
+
+// Two sycl_iterator's (with different access modes)
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+struct is_equality_comparable<sycl_iterator<_Mode1, _T, _Allocator>, sycl_iterator<_Mode2, _T, _Allocator>>
+    : std::true_type
+{
+};
+
+// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-constexpr bool
+struct is_equality_comparable<_Iterator1, _Iterator2,
+                              std::void_t<decltype(::std::decay_t<_Iterator1>() == ::std::decay_t<_Iterator2>())>>
+    : std::true_type
+{
+};
+
+// Check the ability to compare two sycl_iterator's
+/*
+ * @return bool - true, if these iterators may be the same;
+ *         false - we know these iterators are different
+ */
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+bool
+__check_if_iterator_equality_is_possible(sycl_iterator<_Mode1, _T, _Allocator> __it1,
+                                         sycl_iterator<_Mode2, _T, _Allocator> __it2)
+{
+    const auto buf1 = __it1.get_buffer();
+    const auto buf2 = __it2.get_buffer();
+
+    // We are unable to compare two sycl_iterator's from different buffers
+    if (buf1 != buf2)
+        return false;
+
+    // We are unable to compare two sycl_iterator's if one of them is sub_buffer
+    if (buf1.is_sub_buffer() || buf2.is_sub_buffer())
+        return false;
+
+    return true;
+}
+
+template <typename _Iterator1, typename _Iterator2>
+bool
+__check_if_iterator_equality_is_possible(_Iterator1, _Iterator2)
+{
+    return true;
+}
+
+template <typename _Iterator1, typename _Iterator2>
+bool
 __check_equal_iterators(_Iterator1 __it1, _Iterator2 __it2)
 {
-    // In-place exclusive scan works correctly only if an input and an output iterators are the same type.
-    // Otherwise, there is no way to check an in-place case and a workaround below is not applied.
-    if constexpr (::std::is_same_v<::std::decay_t<_Iterator1>, ::std::decay_t<_Iterator2>>)
+    // 1. operator==(_Iterator1, _Iterator2) should be available in compile-time
+    if constexpr (!is_equality_comparable<_Iterator1, _Iterator2>::value)
     {
-        return __it1 == __it2;
+        return false;
```

Comment:
Just because iterators are not equality comparable does not mean they not equal, it means "we don't know".  It seems like this may lead to being too aggressive with the faster algorithm when we are not sure if the algorithm is in-place or not.

I think when they are not comparable, it means they may be equal and we must pick the conservative algorithm.

Reply:
I rather disagree. I think the iterator implementation can, and should, guarantee that any two instances created for the same sequence are comparable. Thus, if instances are not comparable, we consider them being created for different sequences (buffers) and therefore we do know that the algorithm does not run in place.

Reply:
Yes, I understand your point on iterators in general.  I think I was wrong for the general case here.

However, on that note, I think that our `sycl_iterator` implementation breaks this rule when we take sub_buffers into account.

Psuedo-code example:

```
buffer A(n);
buffer B(A, m); // sub_buffer with offset m

sycl_iterator iter1 = begin(A) + m;
sycl_iterator iter2 = begin(B);
sycl_iterator iter3 = begin(B) + m;

iter1 == iter2; // this is will return false, but these point to the same data 
//We only use comparison of idx and don't take a sub_buffer's offset into account

iter1 == iter3; // this is will return true, but these point to different data in the same parent buffer
```
I would argue that neither of these comparisons are valid and these iterators are not comparable because they are based on different buffers.  However, they **can** point to the same underlying data.  I think we can live with this, but we have to be aware of it in this context.

I don't think there is a way to extract the starting offset from a sub_buffer, but I'd be happy to be wrong about that.

Reply:
You are right about the sub-buffers, but the same considerations equally apply to many other algorithms, which might have an incorrect behavior if they are used with overlapping sub-buffers. And I believe such usage violates the requirements of the C++ standard (or at least the expectations) that algorithms apply to non-overlapping sequences or, in case of in-place usage, to exactly the same sequence passed more than one time to the call.

In other words, I believe applying an algorithm to an overlapping sequence, no matter how it is created, results in undefined behavior (i.e. we can assume it does not happen in a correct program). Perhaps it deserves being documented somewhere. 

So, this covers the case of `iter1 != iter2` you described. And for the case of `iter1 == iter3` while they are actually not equal - this is no different than comparing iterators over different buffers, is not it?

Reply:
In short, I think what I said above means that we consider the sequences for sub-buffers over the same buffer being different, and if they actually overlap, passing those to an algorithm would result in undefined behavior.

Reply:
This gets a little tricky, because the `operator==` we are using to compare buffers compares the root buffers.
Therefore, its not easy to determine if we have the same sub-buffer or different sub-buffers.

Reply:
I see.

OK, then what we get in practice? The actual data represented by sub-buffers may match, overlap, or not overlap within the "root" buffer. If they match, iterators obtained with begin() will compare equal, and in-place is correctly detected. If they overlap, it's undefined behavior, so we simply ignore this case. If they do not overlap, iterators will still compare equal, and we use the more pessimistic in-place algorithm - but still everything works correctly. I'd say it is acceptable. In any case, it seems better than refusing to compare iterators if any of them is over a sub-buffer (I'll leave the detailed analysis of that case as an exercise :)). Do we have other ways to deal with sub-buffers?

Reply:
I think we have two options for `sycl_iterators` and `__check_if_iterator_equality_is_possible()`:

1) Only check if the buffers are `operator==`, FALSE: out-of-place, TRUE: in-place
or 
2) Check if the buffers are `operator==` FALSE: out-of-place True: continue
    Check if either buffer is a sub-buffer TRUE: in-place  False: continue
    Make (now valid) check of `sycl_iterator` `operator==`  TRUE: in-place, FALSE: out-of-place

(2) Is more complicated but allows us to catch the one (likely rare) extra case where two `sycl_iterators` are non-overlapping, but both from the same "root" buffer.

I'd be fine with either choice.  However, I don't believe we can know `sycl_iterator` `operator==` is valid unless we confirm that sub-buffers are not involved.  Without validity we can't trust the result in either direction.

Reply:
OK. I think I am fine with the option (2).

Reply:
Fixed: option (2) implemented.